tdiag/src/commands/arrangements.rs

```diff
@@ -0,0 +1,182 @@
+use differential_dataflow::operators::{Count, Join};
+use DifferentialEvent::{Batch, Merge, MergeShortfall};
+
+use tdiag_connect::receive::ReplayWithShutdown;
+
+/// Prints the number of tuples maintained in each arrangement.
+///
+/// 1. Listens to incoming connections from a differential-dataflow
+/// program with timely and differential logging enabled;
+/// 2. runs a differential-dataflow program to track batching and
+/// compaction events and derive number of tuples for each trace;
+/// 3. prints the current size alongside arrangement names;
+pub fn listen(
+    timely_configuration: timely::Configuration,
+    timely_sockets: Vec<Option<std::net::TcpStream>>,
+    differential_sockets: Vec<Option<std::net::TcpStream>>,
+    output_interval_ms: u64, 
+) -> Result<(), crate::DiagError> {
+    let timely_sockets = Arc::new(Mutex::new(timely_sockets));
+    let differential_sockets = Arc::new(Mutex::new(differential_sockets));
+
+    let is_running = Arc::new(std::sync::atomic::AtomicBool::new(true));
+    let is_running_w = is_running.clone();
+
+    timely::execute(timely_configuration, move |worker| {
+        let timely_sockets = timely_sockets.clone();
+        let differential_sockets = differential_sockets.clone();
+
+        let timely_replayer = tdiag_connect::receive::make_readers::<
+            Duration,
+            (Duration, WorkerIdentifier, TimelyEvent),
+        >(
+            tdiag_connect::receive::ReplaySource::Tcp(timely_sockets),
+            worker.index(),
+            worker.peers(),
+        )
+        .expect("failed to open timely tcp readers");
+
+        let differential_replayer = tdiag_connect::receive::make_readers::<
+            Duration,
+            (Duration, WorkerIdentifier, DifferentialEvent),
+        >(
+            tdiag_connect::receive::ReplaySource::Tcp(differential_sockets),
+            worker.index(),
+            worker.peers(),
+        )
+            .expect("failed to open differential tcp readers");
+
+        let window_size = Duration::from_millis(output_interval_ms);
+
+        worker.dataflow::<Duration, _, _>(|scope| {
+            let operates = timely_replayer
+                .replay_with_shutdown_into(scope, is_running_w.clone())
+                .flat_map(|(t, worker, x)| {
+                    if let Operates(event) = x {
+                        Some((
+                            (
+                                (worker, event.id),
+                                format!("{} ({:?})", event.name, event.addr),
+                            ),
+                            t,
+                            1 as isize,
+                        ))
+                    } else {
+                        None
+                    }
+                })
+                .as_collection();
+
+            let events =
+                differential_replayer.replay_with_shutdown_into(scope, is_running_w.clone());
+
+            // @TODO Think about where to put this.
```

Comment:
What does this do? Can we remove it before we merge?

Reply:
The idea here was to log the time spent merging batches. Which maybe belongs more into a differential-aware version of `profile`?